File: doomsday.cpp
#include <iostream>
#include <fstream>
#include <queue> 

using namespace std;

struct pixel { 
    char symbol;
    int t;
    int pi, pj; 
    };


int main (int argc, char **argv) {
    
    

    ifstream myReadFile;

    myReadFile.open(argv[1]);

    char worldmap[1000][1000], temp;
    int i=1, j=1, N, M;

    if (myReadFile.is_open()) {
        temp = myReadFile.get();
        while (temp!=EOF) {
            j=1;
            while (temp!='\n') {
                worldmap[i][j] = temp;
                temp = myReadFile.get();
                j++;
                M=j;
            }
            temp = myReadFile.get();
            i++;
        }
    }

    M++;
    N=i+1;
    
    /*cout << N << ' ' <<M <<endl;*/
    myReadFile.close();

    
    for (i=0;i<N;i++){
        worldmap[i][0]='x';
    }
    for (i=0;i<N;i++){
        worldmap[i][M-1]='x';
    }
    for (j=0;j<M;j++){
        worldmap[0][j]='x';
    }
    for (j=0;j<M;j++){
        worldmap[N-1][j]='x';
    }
        



    int current_t=0;
    bool doomsday=false, change=true;

    queue<pixel> myqueue;
    
    while((doomsday == false) && (change ==true)){
        for (i=1;i<N-1;i++){
            for (j=1;j<M-1;j++){
                if ((worldmap[i][j]=='+')||(worldmap[i][j]=='-')) {
                    pixel temp;
                    temp.symbol=worldmap[i][j];
                    temp.t=current_t;
                    temp.pi=i;
                    temp.pj=j;
                    myqueue.push(temp);
                }	
            }
        }
        change=false;
        
        while(!myqueue.empty()){
            pixel temp = myqueue.front();
            myqueue.pop();
            if(temp.symbol=='+'){
                if(worldmap[temp.pi-1][temp.pj]=='-'){
                    worldmap[temp.pi-1][temp.pj]='*';
                    doomsday=true;}
                else if(worldmap[temp.pi-1][temp.pj]=='.'){
                    worldmap[temp.pi-1][temp.pj]='+';
                    change=true;}

                if(worldmap[temp.pi+1][temp.pj]=='-'){
                    doomsday=true;
                    worldmap[temp.pi+1][temp.pj]='*';}
                else if(worldmap[temp.pi+1][temp.pj]=='.'){
                    worldmap[temp.pi+1][temp.pj]='+';
                    change=true;}

                if (worldmap[temp.pi][temp.pj-1]=='-'){
                    worldmap[temp.pi][temp.pj-1]='*';
                    doomsday=true;}
                else if(worldmap[temp.pi][temp.pj-1]=='.'){
                    worldmap[temp.pi][temp.pj-1]='+';
                    change=true;}

                if (worldmap[temp.pi][temp.pj+1]=='-'){
                    worldmap[temp.pi][temp.pj+1]='*';
                    doomsday=true;}
                else if(worldmap[temp.pi][temp.pj+1]=='.'){
                    worldmap[temp.pi][temp.pj+1]='+';
                    change=true;}
            }
            else {
                if(worldmap[temp.pi-1][temp.pj]=='+'){
                    worldmap[temp.pi-1][temp.pj]='*';
                    doomsday=true;}
                else if(worldmap[temp.pi-1][temp.pj]=='.'){
                    worldmap[temp.pi-1][temp.pj]='-';
                    change=true;}

                if(worldmap[temp.pi+1][temp.pj]=='+'){
                    doomsday=true;
                    worldmap[temp.pi+1][temp.pj]='*';}
                else if(worldmap[temp.pi+1][temp.pj]=='.'){
                    worldmap[temp.pi+1][temp.pj]='-';
                    change=true;}

                if (worldmap[temp.pi][temp.pj-1]=='+'){
                    worldmap[temp.pi][temp.pj-1]='*';
                    doomsday=true;}
                else if(worldmap[temp.pi][temp.pj-1]=='.'){
                    worldmap[temp.pi][temp.pj-1]='-';
                    change=true;}

                if (worldmap[temp.pi][temp.pj+1]=='+'){
                    worldmap[temp.pi][temp.pj+1]='*';
                    doomsday=true;}
                else if(worldmap[temp.pi][temp.pj+1]=='.'){
                    worldmap[temp.pi][temp.pj+1]='-';
                    change=true;}
            }
            
        }
        /*cout<< '\n';
            for (i=1;i<N-1;i++){
                for (j=1;j<M-1;j++){
                    cout<< worldmap[i][j];
                }
            cout<< '\n';
            }*/
        if ((change == false) && (doomsday == false)){
            cout << "the world is saved";
            cout<< '\n';
            for (i=1;i<N-1;i++){
                for (j=1;j<M-1;j++){
                    cout<< worldmap[i][j];
                }
            cout<< '\n';
            }
            
        }
        current_t++;
        if (doomsday == true){
            cout << current_t;
            cout<< '\n';
            for (i=1;i<N-1;i++){
                for (j=1;j<M-1;j++){
                    cout<< worldmap[i][j];
                }
            cout<< '\n';
            }
        }

    }
    
    
    return 0;
}

File: doomsday.py
import re
import sys
import numpy as np
import copy

file_name = sys.argv[1]
fp = open(file_name)
content= fp.readlines()
content = [x.strip() for x in content]
fp.close()
worldmap = list(content)
columns = len(worldmap[0])
rows= len(worldmap)


class element:
    def __init__(self,s,i,j):
        self.symbol = s
        self.x = i
        self.y = j

#class wmap:
 #   def __init__(self,s):
  #      self.symbol = s
   #     self.expand = 0


for i in range(0,rows):
    worldmap[i]='x'+worldmap[i]+'x'

#print(worldmap)

columns=columns+2
rows=rows+2

s = ""
for i in range(0,columns):
    s=s+'x'


worldmap.insert(0,s)
worldmap.append(s)

for i in range(0,rows):
    worldmap[i]=list(worldmap[i])

#print(worldmap)

change=1
doomsday=0
day = 0
backup = list()

while((doomsday==0) and (change==1)):
    change = 0
    day = day +1
    for i in range (0,rows-1):
        for j in range (0,columns-1):
            if ( worldmap[i][j] == '+' or worldmap[i][j] == '-'):
                backup.append(element(worldmap[i][j],i,j))
                
    #print(len(backup))	
    while(len(backup) != 0):
        temp = backup.pop()
        if (temp.symbol == '+'):
            op = '-'
        else:
            op = '+'
        if (worldmap[temp.x+1][temp.y] == op):
            worldmap[temp.x+1][temp.y] ='*' 
            doomsday = 1
        elif (worldmap[temp.x+1][temp.y] == '.'):
            worldmap[temp.x+1][temp.y] =temp.symbol
            change = 1
        if (worldmap[temp.x-1][temp.y] == op):
            worldmap[temp.x-1][temp.y] =  '*' 
            doomsday = 1
        elif (worldmap[temp.x-1][temp.y] == '.'):
            worldmap[temp.x-1][temp.y] =  temp.symbol
            change = 1
        if (worldmap[temp.x][temp.y-1] == op):
            worldmap[temp.x][temp.y-1] = '*'
            doomsday = 1
        elif (worldmap[temp.x][temp.y-1] == '.'):
            worldmap[temp.x][temp.y-1] = temp.symbol
            change = 1
        if (worldmap[temp.x][temp.y+1] == op):
            worldmap[temp.x][temp.y+1] ='*'
            doomsday = 1
        elif (worldmap[temp.x][temp.y+1] == '.'):
            worldmap[temp.x][temp.y+1] =temp.symbol
            change = 1
        
    if ((change==0) and (doomsday==0)):
        print("the world is saved")
        for i in range (1,rows-1):
            for j in range (1,columns-1):
                print(worldmap[i][j],end='')
            print('\n',end='')
    
    if (doomsday == 1):
        print(day)
        for i in range (1,rows-1):
            for j in range (1,columns-1):
                print(worldmap[i][j],end='')
            print('\n',end='')
